core-bundle/src/Resources/contao/modules/ModuleRegistration.php

```diff
@@ -89,6 +89,11 @@ protected function compile()
 			}
 		}
 
+		// Purge expired opt-in tokens (#3709)
+		/** @var OptIn $optIn */
+		$optIn = System::getContainer()->get('contao.opt-in');
+		$optIn->purgeTokens();
```

Comment:
I don't like that this code is now always run in the registration module, because it will affect the performance. Purging tokens should be done in the cron job only.

Reply:
> Purging tokens should be done in the cron job only.

We can't rely on the cron job for purging tokens. No matter how often the cron job runs, there will always be a time period where it is possible that a member can neither login, activate or re-register.

Reply:
The period has been reduced from up to 24 hours to up to 60 minutes. That should suffice for most scenarios. And if someone really needs to further reduce the period, they can adjust the cron job to run minutely, can't they?

Reply:
@leofeyer what about adding 

```php
if (Input::post('FORM_SUBMIT') == $strFormId)
```

then the purging will only be done when the form is actually submitted, ensuring that you will always be able to re-register (if applicable).

Reply:
Or we additionally introduce a `purgeTokensForEmail` function, so that we can target the affected email address specifically. This would then only be one additional database query per regular registration.

```php
if (Input::post('FORM_SUBMIT') == $strFormId && $email == Input::post('email'))
{
    $optIn->purgeTokensForEmail($email);
}
```

Reply:
@contao/developers WDYT?

Reply:
Actually, the opt-in tokens do not matter. We need to make sure that the expired registrations are deleted, when the form is submitted.